internal/servers/worker/internal/metric/proxy_ws.go

```diff
@@ -0,0 +1,123 @@
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "active_connections",
+			Help:      "Count of open websocket connections to Boundary workers.",
+		},
+	)
+
+	bytesReceived = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "received_bytes_total",
+			Help:      "Count of received bytes for Worker connections.",
+		},
+	)
+
+	bytesSent = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: globals.MetricNamespace,
+			Subsystem: proxyWebsocketSubsystem,
+			Name:      "sent_bytes_total",
+			Help:      "Count of sent bytes for Worker connections.",
+		},
+	)
+)
+
+func MustInitializeProxyMetrics(r prometheus.Registerer) {
+	if r == nil {
+		return
+	}
+	r.MustRegister(activeConnCount, bytesReceived, bytesSent)
+}
+
+func ProxyWsMetricsHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		nextRw := rw
+		if rwh, ok := rw.(http.Hijacker); ok {
+			nextRw = &metricsWrapper{ResponseWriter: rw, Hijacker: rwh}
+		}
+		next.ServeHTTP(nextRw, req)
+	})
+}
+
+type wsConn struct {
+	*sync.Once
+	net.Conn
+}
+
+func (wc *wsConn) Read(b []byte) (int, error) {
+	n, err := wc.Conn.Read(b)
+	if err != nil {
+		return n, err
+	}
+	bytesReceived.Add(float64(len(b))) // The slice is always at least 4096 bytes long (`bufio.ReadWriter` default)
+	return n, nil
+}
+
+func (wc *wsConn) Write(b []byte) (int, error) {
+	n, err := wc.Conn.Write(b)
+	if err != nil {
+		return n, err
+	}
+	bytesSent.Add(float64(len(b)))
+	return n, nil
+}
+
+func (wc *wsConn) Close() error {
+	wc.Once.Do(func() { activeConnCount.Dec() })
+	return wc.Conn.Close()
+}
+
+type metricsWrapper struct {
+	// The `Once` objects constrain incrementing/decrementing active connections.
+	// Only a precautionary measure as:
+	// 1. http.(Hijacker).Hijack() cannot be called more than once (2nd call is an error).
+	// 2. websocket.(*Conn).Close() cannot be called more than once (2nd call is no-op).
+	inc sync.Once
+	dec sync.Once
+
+	// `ServeHTTP` takes an `http.ResponseWriter`, so `metricsWrapper`
+	// needs to implement it. Other than that, `ResponseWriter` is unused.
+	http.ResponseWriter
+
+	http.Hijacker
+}
+
+func (mw *metricsWrapper) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	c, rw, err := mw.Hijacker.Hijack()
+	if err != nil {
+		return nil, nil, err
+	}
+	c = &wsConn{Conn: c, Once: &mw.dec}
+
+	// Our underlying `Hijack` call returns a `bufio.ReadWriter` object
+	// with the `Reader` and `Writer` set to the `net.Conn` object *before* we wrap it.
+	// The websocket library uses this buffer to Read and Write. This means
+	// our Conn implementation would never be called.
+	// You might have noticed we're not replacing the Reader. This is
+	// because the Websocket library already does this in `Accept`.
+	rw = bufio.NewReadWriter(rw.Reader, bufio.NewWriter(c))
```

Comment:
I implemented it like this because [this is how it's done in the HTTP library](https://github.com/golang/go/blob/1a955bcdb701d788b048a39d7273621729e257bc/src/net/http/server.go#L310-L329=), so to the extent that it hasn't been a problem until now, I don't foresee this being a problem in the future.

I didn't want to break the pattern just in case I failed to understand some of this. I did try using `c` directly and that also works.

It's probably worth to mention that from my testing `c` is always of type `tls.Conn` since we're always talking via TLS with the Worker. `tls.Conn` has it's own `Write` which [**can** buffer writes at this level](https://github.com/golang/go/blob/master/src/crypto/tls/conn.go#L913-L922=), but it looks like it only does so during handshakes (for both [client](https://github.com/golang/go/blob/master/src/crypto/tls/handshake_client.go#L402=) and [server](https://github.com/golang/go/blob/master/src/crypto/tls/handshake_server.go#L73=) connections). 

The underlying `net.Conn` inside `tls.Conn` seems to just be one of the implementations that come from the `net` package (eg. TCPConn), which seem to eventually end up in `rawConn` methods, which [only write onto the raw connection with no buffering](https://github.com/golang/go/blob/master/src/net/rawconn.go#L51-L61=).

Reply:
Thank you, that helps clarify things a great deal.